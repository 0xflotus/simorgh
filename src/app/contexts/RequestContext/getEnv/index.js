const getEnv = origin => {
  if (
    origin.includes('local') ||
    (process && process.env && process.env.APP_ENV === 'local')
  ) {
    return 'local';
  }

  if (origin.includes('test')) {
    return 'test';
  }

  return 'live';
};

export default getEnv;
