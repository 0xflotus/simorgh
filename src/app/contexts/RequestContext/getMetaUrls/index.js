const base = `https://www.`

const getAmpUrl = url => `${url}.amp`

const getCanonicalUrl = (origin, service, pageType, id) => {
    switch (pageType) {
        case 'frontPage':
            return `${origin}/${service}`;
        default:
            return `${origin}/${service}/articles/${id}`;
    }
}

const getCanonicalUrlWithVariant = (origin, service, serviceVariant, pageType, id) => {
    switch (pageType) {
        case 'frontPage':
            return `${origin}/${service}/${serviceVariant}`;
        default:
            return `${origin}/${service}/articles/${id}/${serviceVariant}`;
    }
}

const getUkConicalUrl = (url) => {
    const tldRegex = /(.com|.co.uk)/g
    return url.replace(tldRegex, '.com')
}

const getNonUkConicalUrl = (url) => {
    const tldRegex = /(.com|.co.uk)/g
    return url.replace(tldRegex, '.co.uk')
}

const getMetaUrls = (origin, env, platform, service, serviceVariant, pageType, id) => {
    let
        canonicalLink = getCanonicalUrl(origin, service, pageType, id),
        canonicalUkLink = getUkConicalUrl(canonicalLink),
        canonicalNonUkLink = getNonUkConicalUrl(canonicalLink);

    if (serviceVariant) {
        canonicalLink = getCanonicalUrlWithVariant(origin, service, serviceVariant, pageType, id);
        return {
            canonicalLink: canonicalLink,
            ampLink: getAmpUrl(canonicalLink),
            canonicalUkLink: canonicalUkLink,
            ampUkLink: getAmpUrl(canonicalUkLink),
            canonicalNonUkLink: canonicalNonUkLink,
            ampNonUkLink: getAmpUrl(canonicalNonUkLink),
        }
    }

    return {
        canonicalLink: canonicalLink,
        ampLink: getAmpUrl(canonicalLink),
        canonicalUkLink: canonicalUkLink,
        ampUkLink: getAmpUrl(canonicalUkLink),
        canonicalNonUkLink: canonicalNonUkLink,
        ampNonUkLink: getAmpUrl(canonicalNonUkLink),
    }
}

export default getMetaUrls