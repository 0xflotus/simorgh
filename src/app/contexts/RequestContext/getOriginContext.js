import get from '../../helpers/json/deepGet';
import onClient from '../../helpers/onClient';

const getArticleID = articleData => {
  const aresID = get(['metadata', 'id'], articleData);
  return aresID ? aresID.split(':').pop() : 'unknown';
};

const getOrigin = bbcOrigin => {
  let origin = 'https://www.bbc.co.uk';

  if (bbcOrigin) {
    origin = bbcOrigin;
  } else if (onClient() && window.location.origin) {
    origin = window.location.origin; // eslint-disable-line prefer-destructuring
  }

  return origin;
};

const getOriginContext = (bbcOrigin, service, articleData) => {
  const origin = getOrigin(bbcOrigin);
  let href = `${origin}/${service}/${getArticleID(articleData)}`;
  let isUK = true;
  let referrer = null;
  let env = 'live';

  if (onClient()) {
    if (window.location.href) {
      href = window.location.href; // eslint-disable-line prefer-destructuring
    }

    if (document.referrer) {
      referrer = document.referrer; // eslint-disable-line prefer-destructuring
    }
  }

  if (origin.includes('.stage.')) {
    env = 'stage';
  }

  if (origin.includes('.test.')) {
    env = 'test';
  }

  if (origin.includes('localhost')) {
    env = 'local';
  }

  if (origin.includes('.com')) {
    isUK = false;
  }

  return {
    origin,
    isUK,
    env,
    href,
    referrer,
  };
};

export default getOriginContext;
