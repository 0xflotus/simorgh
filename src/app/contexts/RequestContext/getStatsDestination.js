const getStatsDestination = ({ isUK = true, env = 'test', service }) => {
  let destination = '';

  const destinationIDs = {
    NEWS_PS: 598285,
    NEWS_GNL: 598287,
    NEWS_PS_TEST: 598286,
    NEWS_GNL_TEST: 598288,
    WS_NEWS_LANGUAGES: 598342,
    WS_NEWS_LANGUAGES_TEST: 598343,
  };

  if (service === 'news') {
    destination = isUK !== false ? 'NEWS_PS' : 'NEWS_GNL';
  } else {
    destination = 'WS_NEWS_LANGUAGES';
  }

  const key = env === 'live' ? destination : `${destination}_TEST`;

  return destinationIDs[key] || destinationIDs.NEWS_PS;
};

export default getStatsDestination;
