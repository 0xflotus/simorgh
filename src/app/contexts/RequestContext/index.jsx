import React from 'react';
import { bool, node, string } from 'prop-types';

export const RequestContext = React.createContext('default');

export const RequestContextProvider = ({
  children,
  env,
  id,
  isUK,
  origin,
  platform,
  statsDestination,
  statsPageIdentifier,
}) => {
  const value = {
    env,
    id,
    isUK,
    origin,
    platform,
    statsDestination,
    statsPageIdentifier,
  };

  return (
    <RequestContext.Provider value={value}>{children}</RequestContext.Provider>
  );
};

RequestContextProvider.propTypes = {
  children: node.isRequired,
  env: string,
  id: string,
  platform: string.isRequired,
  isUK: bool.isRequired,
  origin: string.isRequired,
  statsDestination: string.isRequired,
  statsPageIdentifier: string.isRequired,
};

RequestContextProvider.defaultProps = {
  id: null,
  env: 'local',
};
