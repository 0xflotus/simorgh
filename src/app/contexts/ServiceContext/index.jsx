import React from 'react';
import { node, string } from 'prop-types';
import services from '../../lib/config/services';
import defaultService from '../../lib/config/services/default';
import createLoadableContext from '../utils/createLoadableContext';

/* Create ServiceContext using the default service */
export const ServiceContext = React.createContext(defaultService);

/*
*
* It's important to note that this requires that you declare all of your loadable
* components when modules are initialized rather than when your app is being rendered.
* https://github.com/jamiebuilds/react-loadable#loadablepreloadall
* 
*/
const loadableContexts = {};

Object.keys(services).map(service => {
  loadableContexts[service] = createLoadableContext(
    ServiceContext,
    services[service],
  );
});

export const ServiceContextProvider = ({ children, service }) => {
  const LoadableServiceContextProvider = loadableContexts[service];

  if (!LoadableServiceContextProvider) {
    return null;
  }

  return (
    <LoadableServiceContextProvider>{children}</LoadableServiceContextProvider>
  );
};

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string,
};

ServiceContextProvider.defaultProps = {
  service: 'default',
};
