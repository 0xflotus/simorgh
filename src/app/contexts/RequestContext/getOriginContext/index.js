import { isLocal } from '../getEnv';

const getOriginContext = bbcOrigin => {
  let origin = 'https://www.bbc.co.uk';
  let isUK = true;

  if (bbcOrigin) {
    origin = bbcOrigin;
  } else if (isLocal) {
    // For local dev, set origin to Test
    // Ensures analytics are sent to Test bucket
    origin = 'https://www.test.bbc.co.uk';
  } else if (
    typeof window !== 'undefined' &&
    window.location &&
    window.location.origin
  ) {
    origin = window.location.origin; // eslint-disable-line prefer-destructuring
  }
  if (origin.includes('.com')) {
    isUK = false;
  }

  return {
    origin,
    isUK,
  };
};

export default getOriginContext;
