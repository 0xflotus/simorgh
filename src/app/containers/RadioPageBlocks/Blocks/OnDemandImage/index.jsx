import React, { useContext } from 'react';
import { string } from 'prop-types';
import styled from 'styled-components';
import Image, { AmpImg } from '@bbc/psammead-image';
import {
  GEL_SPACING_QUAD,
  GEL_SPACING_DBL,
} from '@bbc/gel-foundations/spacings';
import { GEL_GROUP_1_SCREEN_WIDTH_MAX } from '@bbc/gel-foundations/breakpoints';
import { RequestContext } from '#contexts/RequestContext';
import { ServiceContext } from '#contexts/ServiceContext';

const paddingDir = ({ dir }) => `padding-${dir === 'rtl' ? 'left' : 'right'}`;

const ImageContainer = styled.div`
  padding: ${GEL_SPACING_QUAD} 0;
  ${paddingDir}: ${GEL_SPACING_DBL};
  @media (max-width: ${GEL_GROUP_1_SCREEN_WIDTH_MAX}) {
    display: none;
  }
`;

const getSrc = ({ imageUrl, size }) =>
  `https://${imageUrl.replace('$recipe', size)}`;
const getSrcSet = imageUrl =>
  `${getSrc({ imageUrl, size: '112x112' })} 112w,${getSrc({
    imageUrl,
    size: '224x224',
  })} 224w`;

const OnDemandImage = ({ imageUrl, dir }) => {
  const { isAmp } = useContext(RequestContext);
  const { defaultImageAltText: alt } = useContext(ServiceContext);
  const height = 224;
  const width = 224;
  const src = getSrc({ imageUrl, size: '112x112' });
  const srcset = getSrcSet(imageUrl);
  const sizes = '(min-width: 1008px) 228px, 30vw';
  const imageProps = { src, alt, srcset };

  return (
    <ImageContainer dir={dir}>
      {isAmp ? (
        <AmpImg
          {...imageProps}
          layout="responsive"
          height={height}
          width={width}
        />
      ) : (
        <Image {...imageProps} sizes={sizes} />
      )}
    </ImageContainer>
  );
};

OnDemandImage.propTypes = {
  imageUrl: string.isRequired,
  dir: string,
};

OnDemandImage.defaultProps = {
  dir: 'ltr',
};

export default OnDemandImage;
