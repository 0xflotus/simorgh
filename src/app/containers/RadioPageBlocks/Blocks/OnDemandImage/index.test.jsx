import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import { render } from '@testing-library/react';
import { ServiceContextProvider } from '#contexts/ServiceContext';
import { RequestContextProvider } from '#contexts/RequestContext';
import OnDemandImage from '.';

// eslint-disable-next-line react/prop-types
const component = ({ url, isAmp, service }) => (
  <ServiceContextProvider service={service}>
    <RequestContextProvider
      isAmp={isAmp}
      service={service}
      pageType="media"
      pathname="/path"
    >
      <OnDemandImage imageUrl={url} />
    </RequestContextProvider>
  </ServiceContextProvider>
);

describe('AudioPlayer blocks OnDemandHeading', () => {
  shouldMatchSnapshot(
    'should render correctly',
    component({ url: 'mock-url', isAmp: false, service: 'news' }),
  );

  it('should ensure the image has the right attributes', () => {
    render(
      component({
        url: 'ichef.bbci.co.uk/images/ic/$recipe/p063j1dv.jpg',
        isAmp: false,
        service: 'pashto',
      }),
    );
    const img = document.querySelector('img');
    expect(img.src).toEqual(
      'https://ichef.bbci.co.uk/images/ic/112x112/p063j1dv.jpg',
    );
    expect(img.alt).toEqual('BBC News پښتو');
    expect(img.srcset).toEqual(
      'https://ichef.bbci.co.uk/images/ic/112x112/p063j1dv.jpg 400w,https://ichef.bbci.co.uk/images/ic/224x224/p063j1dv.jpg 1008w',
    );
  });

  it('should ensure the image has the right attributes for amp', () => {
    render(
      component({
        url: 'ichef.bbci.co.uk/images/ic/$recipe/p063j1dv.jpg',
        isAmp: true,
        service: 'afaanoromoo',
      }),
    );
    const img = document.querySelector('amp-img');
    expect(img.getAttribute('src')).toEqual(
      'https://ichef.bbci.co.uk/images/ic/112x112/p063j1dv.jpg',
    );
    expect(img.getAttribute('alt')).toEqual('BBC News Afaan Oromoo');
    expect(img.getAttribute('layout')).toEqual('responsive');
    expect(img.getAttribute('srcset')).toEqual(
      'https://ichef.bbci.co.uk/images/ic/112x112/p063j1dv.jpg 400w,https://ichef.bbci.co.uk/images/ic/224x224/p063j1dv.jpg 1008w',
    );
  });
});
