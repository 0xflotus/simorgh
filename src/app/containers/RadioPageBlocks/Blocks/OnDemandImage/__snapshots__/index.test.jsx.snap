// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioPlayer blocks OnDemandHeading should render correctly 1`] = `
.c1 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c0 {
  padding: 2rem 0;
  padding-right: 1rem;
}

@media (max-width:24.9375rem) {
  .c0 {
    display: none;
  }
}

<div
  class="c0"
  dir="ltr"
>
  <img
    alt="BBC News"
    class="c1"
    sizes="(min-width: 1007px) 112px, 100vw"
    src="https://mock-url"
    srcset="https://mock-url 112w,https://mock-url 224w"
  />
</div>
`;
